common/src/main/java/org/apereo/openequella/adminconsole/JarService.java

```diff
@@ -0,0 +1,165 @@
+public class JarService {
+	private final String baseUrl;
+	private final File cacheFolder;
+	private final File binFolder;
+
+	public JarService(String baseUrl) {
+		this.baseUrl = baseUrl;
+		this.cacheFolder = Storage.getFolder("cache");
+		this.binFolder = Storage.getFolder("bin");
+	}
+
+	public void ensureBinJars(String... jarNames) {
+		try {
+			for (final String jarName : jarNames) {
+				final File cacheJarFile = ensureCacheJar(jarName);
+				if (cacheJarFile != null) {
+					final File binJarFile = getBinJarFile(jarName);
+					Files.copy(cacheJarFile.toPath(), binJarFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
+				}
+			}
+		} catch (IOException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	/**
+	 * @param jarName Omit the jar extension
+	 */
+	public File ensureCacheJar(String jarName){
+		if (downloadJar(jarName)){
+			return getCacheJarFile(jarName);
+		}
+		return null;
+	}
+
+	/**
+	 * @param jarName Omit the jar extension
+	 * @return true if a jar was downloaded or we already have one available
+	 */
+	public boolean downloadJar(String jarName){
+		boolean isFatal = true;
+		try {
+			//check cache entry
+			final File metadataFile = getCacheMetadataFile(jarName);
+			final File jarFile = getCacheJarFile(jarName);
+			
+			// hit the server to download
+			//final URL url = new URL(baseUrl + "/console.do?jar=" + URLEncoder.encode(jarName + ".jar", "utf-8"));
+			final String strUrl = baseUrl + (baseUrl.endsWith("/") ? "" : "/") + "adminconsole.jar";
+			final URL url = new URL(strUrl);
+			HttpURLConnection conn = null;
+			try {
+				conn = (HttpURLConnection) url.openConnection();
+				conn.setConnectTimeout(10000);
+				conn.setReadTimeout(10000);
+				conn.setRequestMethod("GET");
+				if (metadataFile.exists()){
+					final JarMetadata meta = JsonService.readFile(metadataFile, JarMetadata.class);
+					conn.setRequestProperty("If-None-Match", meta.getEtag());
+					conn.setRequestProperty("If-Modified-Since", meta.getModifiedDate());
+				}
+				conn.connect();
+				final int responseCode = conn.getResponseCode();
+				if (responseCode == 304){
+					// we're ok with our version, but check that it exists first
+					return jarFile.exists();
+				}
+
+				// read etag and modified date
+				final String lastModified = conn.getHeaderField("Last-Modified");
+				final String etag = conn.getHeaderField("ETag");
+				final JarMetadata meta = new JarMetadata();
+				meta.setModifiedDate(lastModified);
+				meta.setEtag(etag);
+				JsonService.writeFile(metadataFile, meta);
+
+				// download and store in our cache
+				try (InputStream is = (InputStream) conn.getInputStream();
+					BufferedInputStream bis = new BufferedInputStream(is);
+					FileOutputStream fos = new FileOutputStream(jarFile)){
+					copyData(bis, fos);
+					return true;
+				}
+			}
+			finally {
+				if (conn != null) {
+					conn.disconnect();
+				}
+			}
+		}
+		catch (Exception e) {
+			if (isFatal){
+				throw new RuntimeException(e);
```

Comment:
it seems isFatal is always true.

Reply:
Yeah. I will remove it.  The idea was that if it failed to download the jar it would use the cached version.  But realistically, if it fails to download the jar you probably won't even be able to contact the Equella server anyway.

Reply:
Have removed the isFatal